etcd-defrag/etcddefrag_controller.go

```diff
@@ -0,0 +1,294 @@
+package etcddefrag
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/go-logr/logr"
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/openshift/hypershift/pkg/etcdcli"
+	"github.com/openshift/hypershift/support/upsert"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+const (
+	pollWaitDuration                       = 2 * time.Second
+	pollTimeoutDuration                    = 60 * time.Second
+	maxDefragFailuresBeforeDegrade         = 3
+	minDefragBytes                 int64   = 100 * 1024 * 1024 // 100MB
+	minDefragWaitDuration                  = 36 * time.Second
+	maxFragmentedPercentage        float64 = 45
+
+	controllerRequeueDuration = 10 * time.Minute
+)
+
+type DefragController struct {
+	client.Client
+	log logr.Logger
+
+	ControllerName string
+	HCPNamespace   string
+	upsert.CreateOrUpdateProvider
+
+	etcdClient         etcdcli.EtcdClient
+	numDefragFailures  int
+	defragWaitDuration time.Duration
+}
+
+type defragTicker struct {
+	defrag *DefragController
+}
+
+func (r *DefragController) setupTicker(mgr manager.Manager) error {
+	ticker := defragTicker{
+		defrag: r,
+	}
+	if err := mgr.Add(&ticker); err != nil {
+		return fmt.Errorf("failed to add defrag ticker runnable to manager: %w", err)
+	}
+	return nil
+}
+
+func (m *defragTicker) Start(ctx context.Context) error {
+	ticker := time.NewTicker(controllerRequeueDuration)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			m.defrag.log.Info("Running defrag.")
+			if err := m.defrag.runDefrag(ctx); err != nil {
+				m.defrag.log.Error(err, "failed to run defragmentation cycle")
+			}
+		}
+	}
+}
+
+func (r *DefragController) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+
+	b := ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.HostedControlPlane{}).
+		WithOptions(controller.Options{
+			RateLimiter: workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 10*time.Second),
+		}).Named(r.ControllerName)
+
+	r.log.Info("NewControllerManagedBy() called")
+	if _, err := b.Build(r); err != nil {
+		return err
+	}
+	endpointsFunc := func() ([]string, error) {
+		return r.etcdEndpoints(ctx)
+	}
+	r.etcdClient = etcdcli.NewEtcdClient(endpointsFunc, events.NewLoggingEventRecorder(r.ControllerName))
+
+	// Set this so that it will immediately requeue itself.
+	r.defragWaitDuration = minDefragWaitDuration
+
+	if err := r.setupTicker(mgr); err != nil {
+		return fmt.Errorf("failed to set up ticker: %w", err)
+	}
+
+	return nil
+}
+
+func (r *DefragController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
the way this is built, why do we need to build the controller and implement the Reconciler interface at all? Couldn't we just let the ticker Runnable be Added in SetupWithManager and that's it?

Reply:
Yes we certainly could.  I'll see about doing this

Reply:
this was resolved but I don't see a clear answer let's unresolve it: do we need to implement the controller runtime interface at all or can we drop this no-op method?

Reply:
Oh sorry! I did this but apparently I didn't get the patch in correctly!  I'll fix it today.

Reply:
Ah, I just didn't remove the Reconcile() function.  Anyway.  Done now.